yarn-project/simulator/src/public/transitional_adaptors.ts

```diff
@@ -91,149 +70,23 @@ export function createPublicExecutionContext(avmContext: AvmContext, calldata: F
     avmContext.persistableState.hostStorage.contractsDb,
     avmContext.persistableState.hostStorage.commitmentsDb,
   );
-
-  return context;
-}
-
-/**
- * Convert the result of an AVM contract call to a PublicExecutionResult for the public kernel
- *
- * @param execution
- * @param newWorldState
- * @param result
- * @returns
- */
-export async function convertAvmResults(
-  executionContext: PublicExecutionContext,
-  newWorldState: JournalData,
-  result: AvmContractCallResults,
-  endMachineState: AvmMachineState,
-): Promise<PublicExecutionResult> {
-  const execution = executionContext.execution;
-
-  const contractStorageReads: ContractStorageRead[] = newWorldState.storageReads.map(
-    read => new ContractStorageRead(read.slot, read.value, read.counter.toNumber(), read.storageAddress),
-  );
-  const contractStorageUpdateRequests: ContractStorageUpdateRequest[] = newWorldState.storageWrites.map(
-    write => new ContractStorageUpdateRequest(write.slot, write.value, write.counter.toNumber(), write.storageAddress),
-  );
-  // We need to write the storage updates to the DB, because that's what the ACVM expects.
```

Comment:
Stream of consciousness trying to make sense of exactly what needs to be written to stateDb....

The same stateDb is used for the entirety of an AVM/ACVM callstack's execution. The ACVM actually updates the stateDb, but the AVM does not.

So, when the ACVM makes a nested call to the AVM, we can be confident that the AVM will be working with the latest stateDb (no need to load in its storage updates to the journal or anything like that). Similarly, when an nested ACVM call returns to the AVM, we don't need to process its storage updates because the stateDb will be updated which is what the AVM uses for its "hostStorage".

We just need to make sure that when a nested call to the AVM returns execution to the ACVM, its pending storage writes need to be committed to the stateDb (since the ACVM doesn't cache writes - or at least does so differently).

Now the question is, when an AVM call returns, what storage writes need to be written to the stateDb? Just that call's or also further nested calls? Maybe whenever an AVM call returns, commit JUST THAT CALL's pending writes to the stateDb (not writes in nested calls / nestedExecutions). The issue with this is that it essentially entirely removes support for reverting side-effects. But maybe that is fine in this transitional phase.

I think this is essentially what you had already @fcarreiro, but now I have the recursive data structure with nestedExecutions and needed to reason about whether I should iterate over just the top level call's writes or also those of all nested calls.

Reply:
Crap but if a nested call to the ACVM updates the stateDb, that stateDb is only the "fallback" from the AVM's perspective. So if we don't ALSO update the AVM's cache, any storage updates that took place PRIOR to the nested call to the ACVM will incorrectly take precedence over the ACVM call's updates :sweat: .... Hence why you also loaded all storage updates from a nested ACVM call into the cache in `updateAvmContext...`.